lib/driver.js

```diff
@@ -383,33 +385,59 @@ class XCUITestDriver extends BaseDriver {
-      this.logEvent('wdaStartAttempted');
-      this.wda = new WebDriverAgent(this.xcodeVersion, this.opts);
+    this.wda = new WebDriverAgent(this.xcodeVersion, this.opts);
 
-      if (this.opts.useNewWDA) {
-        log.debug(`Capability 'useNewWDA' set, so uninstalling WDA before proceeding`);
-        await this.wda.uninstall();
-        this.logEvent('wdaUninstalled');
+    if (this.opts.useNewWDA) {
+      log.debug(`Capability 'useNewWDA' set to true, so uninstalling WDA before proceeding`);
+      await this.wda.quit();
+      await this.wda.uninstall();
+      this.logEvent('wdaUninstalled');
+    } else if (!util.hasValue(this.wda.webDriverAgentUrl)) {
+      log.debug(`Capability 'useNewWDA' set to false, so trying to reuse currently running WDA instance at ${this.wda.url}...`);
+      const noSessionProxy = new NoSessionProxy({
+        server: this.wda.url.hostname,
+        port: this.wda.url.port,
+        base: '',
+        timeout: 3000,
+      });
+      try {
+        const status = await noSessionProxy.command('/status', 'GET');
+        if (!status) {
+          throw new Error(`WDA response to /status command should be defined.`);
+        }
+        log.debug(`Detected WDA listening at ${this.wda.url}. Setting WDA endpoint to ${this.wda.url}`);
+        this.wda.webDriverAgentUrl = this.wda.url;
+      } catch (err) {
+        log.debug(`WDA is not listening at ${this.wda.url}. Rebuilding...`);
       }
+    }
 
-      // local helper for the two places we need to uninstall wda and re-start it
-      let quitAndUninstall = async (msg) => {
-        log.debug(msg);
-        log.debug('Quitting and uninstalling WebDriverAgent, then retrying');
-        await this.wda.quit();
-        await this.wda.uninstall();
-        throw new Error(msg);
-      };
+    // local helper for the two places we need to uninstall wda and re-start it
+    const quitAndUninstall = async (msg) => {
+      log.debug(msg);
+      log.debug('Quitting and uninstalling WebDriverAgent, then retrying');
+      await this.wda.quit();
+      await this.wda.uninstall();
+      throw new Error(msg);
+    };
 
+    const startupRetries = this.opts.wdaStartupRetries || (this.isRealDevice() ? WDA_REAL_DEV_STARTUP_RETRIES : WDA_SIM_STARTUP_RETRIES);
+    const startupRetryInterval = this.opts.wdaStartupRetryInterval || WDA_STARTUP_RETRY_INTERVAL;
+    await retryInterval(startupRetries, startupRetryInterval, async () => {
+      this.logEvent('wdaStartAttempted');
       let wdaStatus = null;
       try {
         wdaStatus = await this.wda.launch(sessionId, realDevice);
       } catch (err) {
         this.logEvent('wdaStartFailed');
-        await quitAndUninstall(`Unable to launch WebDriverAgent because of xcodebuild failure: ${err.message}`);
+        let errorMsg = `Unable to launch WebDriverAgent because of xcodebuild failure: "${err.message}".`;
+        if (this.isRealDevice()) {
+          errorMsg += ` Make sure you follow the tutorial at ` +
```

Comment:
What if I only extract the tutorial URL into a contact? I want to keep message texts separated, since it might be necessary to tune them in the future for launch session and wait for status cases